pca_cca.go

```diff
@@ -0,0 +1,262 @@
+// principal component scores, vecs * a^T, are returned in vars as a
+// min(n,d)-long slice in descending sort order.
+//
+// On failure, vecs and vars are returned nil and ok is returned false.
+func PrincipalComponents(a mat64.Matrix, weights []float64) (vecs *mat64.Dense, vars []float64, ok bool) {
+	n, d := a.Dims()
+	if weights != nil && len(weights) != n {
+		panic("stat: len(weights) != observations")
+	}
+
+	svd, ok := svdFactorizeCentered(a, weights)
+	if !ok {
+		return nil, nil, false
+	}
+
+	vecs = &mat64.Dense{}
+	vecs.VFromSVD(svd)
+	if n < d {
+		// Don't retain columns that are not valid direction vectors.
+		vecs.Clone(vecs.View(0, 0, d, n))
+	}
+	vars = svd.Values(nil)
+	var f float64
+	if weights == nil {
+		f = 1 / float64(n-1)
+	} else {
+		f = 1 / (floats.Sum(weights) - 1)
+	}
+	for i, v := range vars {
+		vars[i] = f * v * v
+	}
+	return vecs, vars, true
+}
+
+// CC is the result of a canonical correlation analysis. A CC should not be used
+// unless it has been returned by a successful call to CanonicalCorrelations.
+type CC struct {
+	// n is the number of observations used to
+	// construct the canonical correlations.
+	n int
+
+	x, y, c *mat64.SVD
+}
+
+// Corrs returns the canonical correlations in dest.
+func (c CC) Corrs(dest []float64) []float64 {
+	if c.c == nil {
+		panic("stat: canonical correlations missing or invalid")
+	}
+
+	return c.c.Values(dest)
+}
+
+// Left returns the left-eigenvectors of the canonical correlation matrix if
+// spheredSpace is true, if spheredSpace is false it returns these eigenvectors
+// back-transformed to the original data space.
+func (c CC) Left(spheredSpace bool) (vecs *mat64.Dense) {
+	if c.c == nil || c.x == nil || c.n < 2 {
+		panic("stat: canonical correlations missing or invalid")
+	}
+
+	var pv mat64.Dense
+	pv.UFromSVD(c.c)
+	if spheredSpace {
+		return &pv
+	}
+
+	var xv mat64.Dense
+	xs := c.x.Values(nil)
+	xv.VFromSVD(c.x)
+
+	scaleColsReciSqrt(&xv, xs)
+
+	pv.Product(&xv, xv.T(), &pv)
+	pv.Scale(math.Sqrt(float64(c.n-1)), &pv)
+	return &pv
+}
+
+// Right returns the right-eigenvectors of the canonical correlation matrix if
+// spheredSpace is true, if spheredSpace is false it returns these eigenvectors
+// back-transformed to the original data space.
+func (c CC) Right(spheredSpace bool) (vecs *mat64.Dense) {
+	if c.c == nil || c.y == nil || c.n < 2 {
+		panic("stat: canonical correlations missing or invalid")
+	}
+
+	var qv mat64.Dense
+	qv.VFromSVD(c.c)
+	if spheredSpace {
+		return &qv
+	}
+
+	var yv mat64.Dense
+	ys := c.y.Values(nil)
+	yv.VFromSVD(c.y)
+
+	scaleColsReciSqrt(&yv, ys)
+
+	qv.Product(&yv, yv.T(), &qv)
+	qv.Scale(math.Sqrt(float64(c.n-1)), &qv)
+	return &qv
+}
+
+// CanonicalCorrelations returns a CC which can provide the results of canonical
+// correlation analysis. The canonical correlation matrix is calculated using
+// the singular value decompositions of the centered input.
```

Comment:
Just a couple of quick thoughts, I'd like to come back to this later (I'm a binge reviewer these days).

How about
```
// CanonicalCorrelations returns a CC which can provide the results of canonical
// correlation analysis of the input data x and y, optionally weighted by weights.
// The canonical correlation matrix is calculated using
// the singular value decompositions of the centered input.
```
I'm not sure whether mentioning SVD is not an implementation detail and also whether the word "centered" does not suggest that the input should already be centered. Perhaps "centered" can be omitted and the rest, if kept, moved towards the end of the comment.